contracts/reverseRegistrar/SignatureReverseResolver.sol

```diff
@@ -0,0 +1,273 @@
+        address addr,
+        string calldata key,
+        string calldata value,
+        uint256 inceptionDate,
+        bytes memory signature
+    )
+        public
+        authorisedSignature(
+            keccak256(
+                abi.encodePacked(
+                    ISignatureReverseResolver
+                        .setTextForAddrWithSignature
+                        .selector,
+                    key,
+                    value
+                )
+            ),
+            addr,
+            inceptionDate,
+            signature
+        )
+        returns (bytes32)
+    {
+        bytes32 node = _getNamehash(addr);
+        _setText(node, key, value, inceptionDate);
+        return node;
+    }
+
+    function _setText(
+        bytes32 node,
+        string calldata key,
+        string calldata value,
+        uint256 inceptionDate
+    ) internal {
+        versionable_texts[recordVersions[node]][node][key] = value;
+        _setLastUpdated(node, inceptionDate);
+        emit TextChanged(node, key, key, value);
+    }
+
+    /**
+     * Returns the text data associated with an ENS node and key.
+     * @param node The ENS node to query.
+     * @param key The text data key to query.
+     * @return The associated text data.
+     */
+    function text(
+        bytes32 node,
+        string calldata key
+    ) public view returns (string memory) {
+        return versionable_texts[recordVersions[node]][node][key];
+    }
+
+    function _setName(
+        bytes32 node,
+        string memory newName,
+        uint256 inceptionDate
+    ) internal virtual {
+        versionable_names[recordVersions[node]][node] = newName;
+        _setLastUpdated(node, inceptionDate);
+        emit NameChanged(node, newName);
+    }
+
+    /**
+     * Returns the name associated with an ENS node, for reverse records.
+     * Defined in EIP181.
+     * @param node The ENS node to query.
+     * @return The associated name.
+     */
+    function name(bytes32 node) public view returns (string memory) {
+        return versionable_names[recordVersions[node]][node];
+    }
+
+    /**
+     * Increments the record version associated with an ENS node.
+     * May only be called by the owner of that node in the ENS registry.
+     * @param addr The node to update.
+     */
+    function _clearRecords(address addr) internal {
+        bytes32 labelHash = LowLevelCallUtils.sha3HexAddress(addr);
+        bytes32 reverseNode = keccak256(
+            abi.encodePacked(parentNode, labelHash)
+        );
+        recordVersions[reverseNode]++;
+        emit VersionChanged(reverseNode, recordVersions[reverseNode]);
+    }
+
+    /**
+     * Increments the record version associated with an ENS node.
+     * May only be called by the owner of that node in the ENS registry.
+     * @param addr The node to update.
+     * @param signature A signature proving ownership of the node.
+     */
+    function clearRecordsWithSignature(
```

Comment:
I am failing to understand why this is only issue for `clearRecords` but not for other `set*` functions.  Also the inclusion of the version doesn't help when the someone keeps omitting to include the `clear` signed message.
Isn't `inceptionDate` enough against most abuses?

Reply:
You're right; to be properly secure, inception dates would need to be recorded on each individual record (in the case of parameterised values such as text, on each key of each record); otherwise someone can replay just the last event, making it impossible to set earlier ones.

If someone omits a message, someone else can supply it - the issue is cases where omitting an earlier message but including a later one makes it impossible to supply the missing message later and arrive at the same end result.

Reply:
`otherwise someone can replay just the last event, making it impossible to set earlier ones.` That would be problematic if we have a counter logic (eg: `n=n+1`) where the state depends on previous state. However, all `setter*` message simply overrides the previous state so there isn't much point supplying the missing message later.

I also didn't fully understand for ` inception dates would need to be recorded on each individual record`

Are you suggesting to change from

```
    function _setLastUpdated(bytes32 node, uint256 inceptionDate) internal {
        lastUpdated[node] = inceptionDate;
    }
```

to 

```
    function _setTextLastUpdated(bytes32 node, string key, uint256 inceptionDate) internal {
        lastUpdated[node][key] = inceptionDate;
    }
```

then change `isAuthorisedWithSignature` to take `key` ?

Reply:
> That would be problematic if we have a counter logic (eg: n=n+1) where the state depends on previous state. However, all setter* message simply overrides the previous state so there isn't much point supplying the missing message later.

They don't, though - setting the address for one coin type invalidates all other previous messages to set addresses for other coin types. The same goes for text records.

Either the last updated value needs to be per-key as you illustrate, or we need to change this to function differently - for example, by taking a complete set of records to set in a single operation, and increment the version at the same time. If going with the former, probably the last updated check should be moved outside `isAuthorisedWithSignature`.

Reply:
Okay, it's getting a bit more complicating than simple fixes. You have been pointing out many changes to the functions not related to this PR but introduced at https://github.com/ensdomains/ens-contracts/pull/265 (you only noticed because I moved the code to different file).

May I suggest to leave it for the separate PR by @jefflau